[WFLY-20512]: Expose Artemis commit-interval attribute for scaledown.

ehsavoie opened the pull request:
* Adding a new attribute commit-interval for scale-down to avoid loading all the messages to be transfered in a single transaction.

Jira: https://issues.redhat.com/browse/WFLY-20512
      https://issues.redhat.com/browse/WFLY-19752

Proposal: https://github.com/wildfly/wildfly-proposals/pull/710
RFE: https://github.com/wildfly/wildfly-proposals/issues/709

rhusar commented on messaging-activemq/subsystem/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties:
```diff
@@ -309,6 +309,7 @@ ha-policy.request-backup=If true then the server will request a backup on anothe
 ha-policy.restart-backup=Will this server, if a backup, restart once it has been stopped because of failback or scaling down.
 ha-policy.scale-down-cluster-name=Name of the cluster used to scale down.
 ha-policy.scale-down-connectors=List of connectors used to form the scale-down cluster.
+ha-policy.scale-down-commit-interval=The batch size for committing the messagesbeing moved on scale down.
```
Missing space in 'messagesbeing'.

marcosgopen commented on messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ha/ScaleDownAttributes.java:
```diff
@@ -112,6 +122,10 @@ static ScaleDownConfiguration addScaleDownConfiguration(OperationContext context
         }
+        ModelNode commitInterval = SCALE_DOWN_COMMIT_INTERVAL.resolveModelAttribute(context, model);
+        if (commitInterval.isDefined()) {
+            scaleDownConfiguration.setCommitInterval(commitInterval.asInt());
```
I would handle here the 'IllegalArgumentException' that can be thrown by asInt() with a default

ehsavoie replied:
It should have been validated but the attribute validation

marcosgopen replied:
I see, thanks @ehsavoie !

yersan commented on messaging-activemq/subsystem/src/test/java/org/wildfly/extension/messaging/activemq/MessagingActiveMQSubsystem_17_0_TestCase.java:
```diff
@@ -0,0 +1,238 @@
+        testTransformers(ModelTestControllerVersion.MASTER, MessagingExtension.VERSION_13_0_0);
+    }
+
+    @Test
+    public void testTransformersEAP_7_4_0() throws Exception {
+        testTransformers(EAP_7_4_0, MessagingExtension.VERSION_13_0_0);
+    }
```
You will also need to test the transformation against EAP_8_0_0, and, once https://github.com/wildfly/wildfly-legacy-test/pull/40 gets merged and https://issues.redhat.com/browse/WFCORE-7254 gets resolved, against EAP_8_1_0

yersan replied:
@ehsavoie This is still an open question; you have not added any transformation against EAP 8.1, so can we assume you prefer to leave it as it is?

bstansberry commented:
Next time please use two commits, one with the boilerplate for the new version and xsd and the other with the functional change.

bstansberry commented on messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/MessagingSubsystemParser_17_0.java:
```diff
@@ -0,0 +1,688 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Please use the standard WildFly header.

bstansberry commented on messaging-activemq/subsystem/src/test/java/org/wildfly/extension/messaging/activemq/MessagingActiveMQSubsystem_17_0_TestCase.java:
```diff
@@ -0,0 +1,238 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Please use the standard WildFly header.

rhusar commented:
> Next time please use two commits, one with the boilerplate for the new version and xsd and the other with the functional change.

@ehsavoie Yes, because now, reviewing requires local checkout to see what actually changed in the XSD.

ehsavoie committed:
[WFLY-20512]: Expose Artemis commit-interval attribute for scaledown.

* Upgrading messaging schema to 17.0

Jira: https://issues.redhat.com/browse/WFLY-20512
      https://issues.redhat.com/browse/WFLY-19752

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

bstansberry commented on messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/ha/ScaleDownAttributes.java:
```diff
@@ -112,6 +122,10 @@ static ScaleDownConfiguration addScaleDownConfiguration(OperationContext context
         }
+        ModelNode commitInterval = SCALE_DOWN_COMMIT_INTERVAL.resolveModelAttribute(context, model);
+        if (commitInterval.isDefined()) {
+            scaleDownConfiguration.setCommitInterval(commitInterval.asInt());
```
Should this AttributeDefinition have a default value?

I was wondering what our typical handling of this kind of thing was, i.e. an argument for having one (or not) would be consistency with the rest of the subsystem.

I don't know if this means we typically have defaults, but this IDE search shows places where maybe the code is trying to deal with -1.  (BTW, should this attribute have this validator?)

```
Interface
    InfiniteOrPositiveValidators
Usages in Project and Libraries  (67 usages found)
    Class static member access  (60 usages found)
        wildfly-messaging-activemq-subsystem  (60 usages found)
            org.wildfly.extension.messaging.activemq  (29 usages found)
                BridgeDefinition  (6 usages found)
                    71 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    72 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    89 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    90 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    143 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    144 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                ClusterConnectionDefinition  (7 usages found)
                    91 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    133 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    134 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    189 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    190 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    212 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    213 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                CommonAttributes  (4 usages found)
                    79 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    103 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    104 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    115 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                GroupingHandlerDefinition  (1 usage found)
                    58 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                ServerDefinition  (11 usages found)
                    127 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    128 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    160 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    161 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    398 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    399 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    474 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    562 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    563 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    587 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    599 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
            org.wildfly.extension.messaging.activemq.ha  (2 usages found)
                HAAttributes  (2 usages found)
                    56 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    57 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
            org.wildfly.extension.messaging.activemq.jms  (12 usages found)
                ConnectionFactoryAttributes.Common  (12 usages found)
                    116 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    138 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    139 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    160 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    182 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    183 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    327 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    328 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    386 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    387 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    398 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    399 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
            org.wildfly.extension.messaging.activemq.jms.bridge  (5 usages found)
                JMSBridgeDefinition  (5 usages found)
                    153 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    158 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    159 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    170 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    171 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
            org.wildfly.extension.messaging.activemq.jms.legacy  (12 usages found)
                LegacyConnectionFactoryDefinition  (12 usages found)
                    105 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    127 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    128 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    149 .setValidator(InfiniteOrPositiveValidators.LONG_INSTANCE)
                    162 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    163 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    288 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    289 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    342 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    343 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
                    354 .setCorrector(InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR)
                    355 .setValidator(InfiniteOrPositiveValidators.INT_INSTANCE)
    Usage in import  (7 usages found)
        wildfly-messaging-activemq-subsystem  (7 usages found)
            org.wildfly.extension.messaging.activemq  (3 usages found)
                ServerDefinition  (3 usages found)
                    18 import static org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators.INT_INSTANCE;
                    19 import static org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators.NEGATIVE_VALUE_CORRECTOR;
                    48 import static org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators.LONG_INSTANCE;
            org.wildfly.extension.messaging.activemq.ha  (1 usage found)
                HAAttributes  (1 usage found)
                    23 import org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators;
            org.wildfly.extension.messaging.activemq.jms  (1 usage found)
                ConnectionFactoryAttributes  (1 usage found)
                    52 import org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators;
            org.wildfly.extension.messaging.activemq.jms.bridge  (1 usage found)
                JMSBridgeDefinition  (1 usage found)
                    44 import org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators;
            org.wildfly.extension.messaging.activemq.jms.legacy  (1 usage found)
                LegacyConnectionFactoryDefinition  (1 usage found)
                    38 import org.wildfly.extension.messaging.activemq.InfiniteOrPositiveValidators;
```

bstansberry replied:
Sounds good.

The 'commitInterval' ModelNode and the isDefined() check are no longer needed.

yersan commented on messaging-activemq/subsystem/src/test/resources/org/wildfly/extension/messaging/activemq/subsystem_17_0_reject_transform.xml:
```diff
@@ -0,0 +1,287 @@
+<subsystem xmlns="urn:jboss:domain:messaging-activemq:16.0">
```
This configuration should include the `commit-interval` attribute, and the test should assert that those resources were rejected; otherwise, we will be only testing the default values, which I understand are already tested using subsystem_17_0_transform.xml

yersan commented on messaging-activemq/subsystem/src/test/resources/org/wildfly/extension/messaging/activemq/subsystem_17_0_transform.xml:
```diff
@@ -0,0 +1,429 @@
+<subsystem xmlns="urn:jboss:domain:messaging-activemq:16.0">
```
This should be using 17.0 version

ehsavoie commented:
Should be all good, @yersan may you take a look please ?

ehsavoie committed:
[WFLY-20512]: Expose Artemis commit-interval attribute for scaledown.

* Adding a new attribute commit-interval for scale-down to avoid loading
  all the messages to be transfered in a single transaction.

Jira: https://issues.redhat.com/browse/WFLY-20512
      https://issues.redhat.com/browse/WFLY-19752

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>

yersan commented on messaging-activemq/subsystem/src/main/java/org/wildfly/extension/messaging/activemq/MessagingTransformerRegistration.java:
```diff
@@ -58,12 +60,19 @@ public String getSubsystemName() {
     @Override
     public void registerTransformers(SubsystemTransformerRegistration registration) {
         ChainedTransformationDescriptionBuilder builder = TransformationDescriptionBuilder.Factory.createChainedSubystemInstance(registration.getCurrentSubsystemVersion());
+        registerTransformers_WF_36(builder.createBuilder(MessagingExtension.VERSION_17_0_0, MessagingExtension.VERSION_16_0_0));
```
nitpick: this method would be more correct if it is named as `registerTransformers_WF_35` instead, since the transformer chain here is against VERSION_16_0_0, which was shipped in WF35

rhusar reviewed (approved):
No more comments from me.

Nevertheless, but I have to mention that we should rethink how we do things in this subsystem especially around classes such as`MessagingSubsystemParser_17_0.java` which are mostly copies of each other as it's not sustainable having to add and maintain excess of 3600 LOC just to support a single new attribute.
